---
title: "Report: Body-sites microbiome in animal model"
subtitle: "Fig4. Drug Repurposing: effect of TB treatment on body-sites"
author: "Aurea Simon-Soro"
date: \today
output:
  pdf_document: default
  html_document: default
---

```{r knitr setup, echo=FALSE}
### ================
###   knitr setup
### ================
library(knitr)
opts_chunk$set(
  tidy=FALSE,
  cache=TRUE,
  echo=FALSE,
  warning=FALSE,
  message=FALSE,
  dpi=100,
  fig.width=8,
  fig.height=8,
  fig.align = "center"
  )
```


```{r R packages, message=FALSE}
### ================
###   R packages
### ================
#This package will also help us more easily manipulate our data
library(magrittr)
library(qiimer)
library(pander)
#Analyses of Phylogenetics and Evolution package. Required for tree calculations to be used with phyloseq
library(ape)
#The vegan package provides tools for descriptive community ecology. It has most basic functions of diversity analysis, community ordination and dissimilarity analysis. In general, this package is used for Bray-Curtis and Jaccard analyses.
library(vegan)
#This package is used to calculate and plot Venn diagrams as well as heatmaps
library(gplots)
library(pheatmap)
#This package will help us more easily manipulate our data, which are matrices
library(usedist)
library(tibble)
#The phyloseq package seeks to address issues with multiple microbiome analysis packages by providing a set of functions that internally manage the organizing, linking, storing, and analyzing of phylogenetic sequencing data. In general, this package is used for UniFrac analyses.
library(phyloseq)
library(tidyverse)
library(ggbeeswarm)
library(ggtern)
library(ggthemes)
library(reshape2)
library(kylemisc)
```



<!-- resources -->
```{r resources}
### ================
###   R resources
### ================
source("/home/aurea/R_functions.R")
```


```{r}
### ===========================
###   define constants
### ===========================
### minimum reads threshold
min_reads <- 1000
### rarefying subsample size 
richness_subsample_size <- 1000
### number of samples threshold to show heatmap on the page
sample_threshold <- 100
### setwd
#fill in your project dir
root_dir = "/home/aurea/Koo/Shared.Species.BODYSITE/Input.Files"
### mapping file path
mapping_file_fp <- file.path(root_dir, "MF.Aurea.txt")
### otu table file path
feature_table_fp <- file.path(root_dir, "feature-table.tsv")
### taxonomic assignment 
taxo_assignment_fp <- file.path(root_dir, "taxonomy.tsv")
### Bray-Curtis Dissimilarities file path
bc_fp <- file.path(root_dir, "bc.distance-matrix.tsv")
### faith
faith_fp <- file.path(root_dir, "alpha-diversity.tsv")
```

```{r, warning=F}
### ===========================
###   read in data
### ===========================
### read mapping file
s <- read_qiime_mapping_file(mapping_file_fp) 
### check for the column names to assign color_by and shape_by for pcoa plots
color_by <- NULL
shape_by <- NULL
potential_headers <- c("SampleType", "study_group", "cage_number")
header_idx = which(is.element(potential_headers, colnames(s)))
if(length(header_idx)>0) {
  color_by <- potential_headers[header_idx[1]]
}
if(length(header_idx)>1) {
  shape_by <- potential_headers[header_idx[2]]
}
### read otu table
counts <- readr::read_delim(feature_table_fp, skip=1, delim="\t") %>%
  column_to_rownames(var = "#OTU ID") %>%
  as.matrix()
### get read counts
read_counts <- colSums(counts) %>% 
  as.data.frame() %>%
  setNames(c("Read_Counts")) %>%
  rownames_to_column(var="SampleID")
### find the samples to keep
s <- merge(s, read_counts, by="SampleID", all.x=T) %>%
  mutate(Keep = Read_Counts > min_reads) %>%
  mutate(isControl = grepl("geneblock|freewater|extract", SampleID, ignore.case = TRUE))
#trun_taxon perl expression is to get rid of trailing "; s__" since it doesn't add any new information
### taxonomic assignment
ta <- read_delim(file=taxo_assignment_fp, delim="\t") %>%
  mutate(trunc_taxon = sub('(; [kpcofgs]__)+$', "", Taxon, perl=T)) %>%
  arrange(order(match(rownames(counts), `Feature ID`)))
### check if the order of the assignments and the order of featue table is the same
if (!all(rownames(counts) == ta$`Feature ID`)) {
  stop (simpleError("The order of the features in the table and classifications don't match"))
}
adf <- split_assignments(ta$trunc_taxon) 
### remove contamination
is_mitochondrial <- grepl("mitochondria", adf$Family)
is_chloroplast <- grepl("Chloroplast", adf$Class)
is_unassigned <- grepl("Unassigned", adf$Kingdom)
is_archaea <- grepl("Archaea", adf$Kingdom)
is_contam <- is_mitochondrial | is_chloroplast | is_unassigned ### Archaea kept to check positive control samples
counts <- counts[!is_contam,]
adf <- adf[!is_contam,]
ta <- ta[!is_contam,]
rm(is_contam, is_mitochondrial, is_chloroplast, is_unassigned, is_archaea)
a <- simplify_assignments(adf, rank1="Phylum", rank2="Genus")
names(a) <- ta$`Feature ID`
summed_cts <- rowsum(counts, a) 
summed_props <- sweep(summed_cts, 2, colSums(summed_cts), "/")
#if we want all the otus
 all_otus <- counts %>%
   merge(ta[,c("Feature ID","Confidence","trunc_taxon")], by.x="row.names", by.y="Feature ID", all.x=T) %>%
   dplyr::rename(Taxon = trunc_taxon) %>%
   dplyr::rename(`Feature ID` = Row.names)
# 
# write.csv(x = all_otus, file = "all_otus.csv", row.names = F)
```


```{r, Samples error check 1}
### ===========================
###   check for missing samples
### ===========================
### possible issue 1: Samples found in the sample sheet but not in the feature table (0 reads)
s_missing <- s %>%
  filter(!SampleID %in% colnames(counts)) %>%
  select(SampleID, SampleType, isControl)
if (any(!s_missing$isControl)) {
  pander(filter(s_missing, !isControl), caption="These samples were in the sample sheet but not in the feature table.")
}
```


```{r, Samples error check 2}
### possible issue 2: Samples found in the feature table but not in the sample sheet. There must be an error!
in_counts_not_in_s <- setdiff(colnames(counts), s$SampleID)
if (length(in_counts_not_in_s) > 0) {
  stop (simpleError("These SampleID(s) are in the feature table, but not found in the sample sheet.", paste(in_counts_not_in_s, collapse=" ")))
}
s[s$SampleID %in% s_missing$SampleID, "Keep"] <- FALSE
```


```{r theme, include=FALSE}
theme_set(theme_classic() + theme(
axis.line.x = element_line(colour = 'black', size=0.5, linetype='solid'),
axis.line.y = element_line(colour = 'black', size=0.5, linetype='solid'))) 

theme_boxplot <- function () {theme_base(base_size = 15) + 
                theme(line = element_line(size=0.8),
                text = element_text(size = 15),
                axis.text.x = element_text(color="black"),
                axis.text.y = element_text(color="black"),
                plot.background=element_blank())}

theme_figures <- function () {theme_kyle() +
                theme(line = element_line(size=0.8),
                text = element_text(size = 15),
                axis.text.x = element_text(color="black"),
                axis.text.y = element_text(color="black"),
                plot.background=element_blank())}
```


```{r general quantification, include=FALSE}
## Whole samples that are above the `r min_reads` read count threshold
pander(table(s[, color_by], factor(ifelse(s$Keep, "Keep", "Discard"))))

#Subset samples geneBlock_DNA
s <- filter(s, !is.na(HostSpecies))

#Subset for the TXT of interest with sucrose as water
s <- filter (s, Water=="Sucrose") %>%
  filter(study_group %in% c("PBS", "TB")) 

```


```{r arrange_and_counts, include=FALSE}

#Re-order categories
s$study_group <- reorder(s$study_group, new.order = c("PBS","TB"))
s$SampleType <- reorder(s$SampleType, new.order = c("Oral Swab", "Dental Plaque", "Feces"))


```

```{r Color Palette, include=FALSE}
cols_group <- c("PBS" ="#ffac57", "TB"= "#c667ef")

cols_Site <- c("Dental Plaque"= "#4743c7", "Oral Swab"= "#6ad54a", "Feces"= "#ae1c00")

```


##Alpha-diversity
# Taxonomic heatmap
```{r}
prop_cut <- 0.01
satu_limit <- 0.4
heatmap_fp <- "otu_heatmap.pdf"
show.text <- sum(s$Keep) > sample_threshold
```

Each column of the heatmap represents one sample and each row represents one taxon, typically a genus. Taxa were included in the chart if the abundance in any sample exceeded `r 100*prop_cut`%. 

The chart is colored white if taxa were not observed in the sample, dark blue if taxa were observed at very low abundance. This allows the reader to quickly survey presence/absence. Abundance values exceeding `r 100*satu_limit`% are colored red, indicating an extremely dominant species.

`r if(show.text){paste0("Please see attached plot ", heatmap_fp, ".")}`

```{r, fig.height=12, fig.width=16}

props_toPlot <- summed_props[, s$SampleID]  
grps <- c(color_by, shape_by)
### grouped heatmap
if (dim(s)[1] > sample_threshold) {
  heatmap_grouped(props_toPlot, s, grps=grps, thre=0.01, option=2, prop_cut = prop_cut, satu_limit=satu_limit, fname = heatmap_fp)
} else {
  heatmap_grouped(props_toPlot, s, grps=grps, thre=0.01, option=2, prop_cut = prop_cut, satu_limit=satu_limit)
}

```

*Figure 1. Heatmap with the bacterial composition and abundances.** This plot shows the bacterial composition of the samples. Sample arrangement corresponds to sample type (dental plaque, oral swab and  feces) and study groups (treatments). We used ASV and GreenGenes database for taxonomy classification.
\newpage


```{r DF Bacteria Proportions Melted}
taxon_props <- props_toPlot[rowMeans(props_toPlot) > 0.01,]
s_taxa <- t(taxon_props)[s$SampleID,] 

set.seed(2021)
pc_df<- cbind(s, s_taxa)

#Create DF with the Bacteria in a single columns for all study_groups for all SampleTypes
bact.prop <- melt(pc_df,id.vars=c('SubjectID', 'SampleType', 'study_group'), measure.vars=c(56:70)) 

```

#Bacteria with more than 1% mean relative abundance
```{r Bar Plot Taxa, fig.height=5}
ggplot(bact.prop, aes(x=study_group, y=value)) +
 geom_col(aes(fill = variable), position="fill", stat = "identity") +
 labs(x= "Treatment", y= "Bacterial Abundance") +
   ggsci::scale_fill_d3(palette = c("category20")) +
  labs(x="", y="Relative Abundance", fill="Bacteria") +
  theme_boxplot() + 
  facet_grid(~SampleType, scales= "free") +
  theme(axis.text.x=element_text(angle=-50, hjust= .1))
```


# Beta diversity
## Bray-Curtis distances
```{r}
dist_fp <- bc_fp
dist_name <- "Bray-Curtis Unifrac distances"
```


###Oral Swab
```{r Obtain the proportiona and axes for the amin genera}
s.partial <- s %>%
  filter(Keep) %>%
  filter(!isControl)%>%
  filter(SampleType=="Oral Swab")

# Make sure the distance matrix is in the same order as the sample table
dist_in <- usedist::dist_subset(read_qiime_distmat(bc_fp), s.partial$SampleID)
# Here is the actual principal coordinates analysis
pc <- pcoa(dist_in)
# Get the first 3 principal coordinates as a matrix,
#   make sure rows are in the same order as sample table
s_coords <- pc$vectors[s.partial$SampleID,1:3]

# Stick the coordinates and taxon proportions onto
#   the right side of the sample table
pc_df <- cbind(s.partial, s_coords)

```

```{r bc Swab}
library(ggrepel)
p1 <- ggplot(pc_df) +
  geom_point(aes(x=Axis.1, y=Axis.2, color= study_group), shape= 21, fill="#ebcdc1", size = 2.5, stroke=2, alpha=0.8) +
 scale_color_manual(values= cols_group) +
   labs(x=paste0("PC1 (", pctvar[1], "%)"),  y=paste0("PC2 (", pctvar[2], "%)"), color="Treatment") + 
  theme_boxplot() +
  #theme(legend.position = "none") + 
  ggtitle("Oral Swab") +
  theme(plot.title = element_text(hjust = 0.5))

```

###Dental Plaque
```{r Obtain the proportiona and axes for the amin genera}
s.partial <- s %>%
  filter(Keep) %>%
  filter(!isControl)%>%
  filter(SampleType=="Dental Plaque")

# Make sure the distance matrix is in the same order as the sample table
dist_in <- usedist::dist_subset(read_qiime_distmat(bc_fp), s.partial$SampleID)
# Here is the actual principal coordinates analysis
pc <- pcoa(dist_in)
# Get the first 3 principal coordinates as a matrix,
#   make sure rows are in the same order as sample table
s_coords <- pc$vectors[s.partial$SampleID,1:3]

# Stick the coordinates and taxon proportions onto
#   the right side of the sample table
pc_df <- cbind(s.partial, s_coords)

```

```{r bc Plaque}
library(ggrepel)
p2<- ggplot(pc_df) +
  geom_point(aes(x=Axis.1, y=Axis.2, color= study_group), shape= 21, fill="#ebcdc1", size = 2.5, stroke=2, alpha=0.8) +
 scale_color_manual(values= cols_group) +
   labs(x=paste0("PC1 (", pctvar[1], "%)"),  y=paste0("PC2 (", pctvar[2], "%)"), color="Treatment") + 
  theme_boxplot() +
  #theme(legend.position = "none") + 
  ggtitle("Dental Plaque") +
  theme(plot.title = element_text(hjust = 0.5))

```

###Feces
```{r Obtain the proportiona and axes for the amin genera}
s.partial <- s %>%
  filter(Keep) %>%
  filter(!isControl)%>%
  filter(SampleType=="Feces")

# Make sure the distance matrix is in the same order as the sample table
dist_in <- usedist::dist_subset(read_qiime_distmat(bc_fp), s.partial$SampleID)
# Here is the actual principal coordinates analysis
pc <- pcoa(dist_in)
# Get the first 3 principal coordinates as a matrix,
#   make sure rows are in the same order as sample table
s_coords <- pc$vectors[s.partial$SampleID,1:3]

# Stick the coordinates and taxon proportions onto
#   the right side of the sample table
pc_df <- cbind(s.partial, s_coords)

```

```{r bc Feces}
library(ggrepel)
p3<- ggplot(pc_df) +
  geom_point(aes(x=Axis.1, y=Axis.2, color= study_group), shape= 21, fill="#ebcdc1", size = 2.5, stroke=2, alpha=0.8) +
 scale_color_manual(values= cols_group) +
   labs(x=paste0("PC1 (", pctvar[1], "%)"),  y=paste0("PC2 (", pctvar[2], "%)"), color="Treatment") + 
  theme_boxplot() +
  #theme(legend.position = "none") + 
  ggtitle("Feces") +
  theme(plot.title = element_text(hjust = 0.5))

```

```{r}
ggpubr::ggarrange(p1, p2, p3, ncol = 3, nrow = 1, common.legend = TRUE)
```

