---
title: "Report: Body-sites microbiome in animal model"
subtitle: "Figure 1: Body-sites microbiome description"
author: "Aurea Simon-Soro"
date: \today
output:
  pdf_document: default
  html_document: default
---

<!-- ================================================================================================ -->
<!--   Beginning of Preamble : Preamble seldom requires change                                        -->
<!-- ================================================================================================ -->

<!-- Notes -->
```{r eval=FALSE, include=FALSE}
#notes
#neat: you can run this following command in the console to give your reports custom names (or date-stamp them)
#rmarkdown::render('ronen_stein_Report_shotgun.Rmd',output_file = paste('CEASE.report.', Sys.Date(), '.pdf', sep=''))
```
<!-- ===== -->

<!-- knitr setup -->
```{r knitr setup, echo=FALSE}
### ================
###   knitr setup
### ================
library(knitr)
opts_chunk$set(
  tidy=FALSE,
  cache=TRUE,
  echo=FALSE,
  warning=FALSE,
  message=FALSE,
  dpi=100,
  fig.width=8,
  fig.height=8,
  fig.align = "center"
  )
```

<!-- R packages -->
```{r R packages, message=FALSE}
### ================
###   R packages
### ================
#This package will also help us more easily manipulate our data
library(magrittr)
library(qiimer)
library(pander)
#Analyses of Phylogenetics and Evolution package. Required for tree calculations to be used with phyloseq
library(ape)
#The vegan package provides tools for descriptive community ecology. It has most basic functions of diversity analysis, community ordination and dissimilarity analysis. In general, this package is used for Bray-Curtis and Jaccard analyses.
library(vegan)
library(usedist)
library(tidyverse)
library(tibble)
#The phyloseq package seeks to address issues with multiple microbiome analysis packages by providing a set of functions that internally manage the organizing, linking, storing, and analyzing of phylogenetic sequencing data. In general, this package is used for UniFrac analyses.
library(phyloseq)
#Pretty Venn disgrams
library(VennDiagram)
library(ggbeeswarm)
library(ggtern)
library(ggthemes)
library(reshape2)
library(kylemisc)
library(ggrepel)
library(ggsci)
```



<!-- resources -->
```{r resources}
### ================
###   R resources
### ================
source("R_functions.R")
```


```{r}
### ===========================
###   define constants
### ===========================
### minimum reads threshold
min_reads <- 1000
### rarefying subsample size 
richness_subsample_size <- 1000
### number of samples threshold to show heatmap on the page
sample_threshold <- 100
### setwd
#fill in your project dir
# From Kyle:
# To keep the project self-contained, place the data inside the project
# directory and use relative paths.  Here, I put the files in inst/extdata.
# If this project ever turns into an R package, you can keep the data files
# right where they are.
root_dir <- "inst/extdata"
### mapping file path
mapping_file_fp <- file.path(root_dir, "MF.Aurea.txt")
### otu table file path
feature_table_fp <- file.path(root_dir, "feature-table.tsv")
### taxonomic assignment 
taxo_assignment_fp <- file.path(root_dir, "taxonomy.tsv")
### Bray-Curtis Dissimilarities file path
bc_fp <- file.path(root_dir, "bc.distance-matrix.tsv")
### faith
faith_fp <- file.path(root_dir, "alpha-diversity.tsv")
```

```{r}
study_group_levels <- c(
  "Control", "IONPs", "GOx", "CHX",
  "IONPs-GOx", "PBS", "NPC", "NPC.Farnesol",
  "Farnesol", "NPC.TB", "TB")
water_levels <- c("Sucrose", "Glucose")
sample_type_levels <- c("Oral Swab", "Dental Plaque", "Feces")
```


```{r, warning=F, message=F}
### ===========================
###   read in data
### ===========================
### read mapping file
s <- read_tsv(mapping_file_fp) %>%
  rename(SampleID = "#SampleID") %>%

  # Fix sequencing control samples
  mutate(SampleType = if_else(
    (study_group %in% "Contamination Control"), "Blank oral swab", SampleType)) %>%
  mutate(is_seq_control = is.na(HostSpecies)) %>%
  mutate(study_color = na_if(study_color, "Contamination Control")) %>%
  mutate(study_group = na_if(study_group, "Contamination Control")) %>%

  # Assign levels
  mutate(study_group = fct_relevel(study_group, study_group_levels)) %>%
  mutate(Water = fct_relevel(Water, water_levels)) %>%
  mutate(SampleType = fct_relevel(SampleType, sample_type_levels))


### read otu table
counts <- readr::read_delim(feature_table_fp, skip=1, delim="\t") %>%
  column_to_rownames(var = "#OTU ID") %>%
  as.matrix()
### get read counts
read_counts <- colSums(counts) %>% 
  as.data.frame() %>%
  setNames(c("Read_Counts")) %>%
  rownames_to_column(var="SampleID")
### find the samples to keep
s <- merge(s, read_counts, by="SampleID", all.x=T) %>%
  mutate(Keep = Read_Counts > min_reads) %>%
  mutate(isControl = grepl("geneblock|freewater|extract", SampleID, ignore.case = TRUE))
#trun_taxon perl expression is to get rid of trailing "; s__" since it doesn't add any new information
### taxonomic assignment
ta <- read_delim(file=taxo_assignment_fp, delim="\t") %>%
  mutate(trunc_taxon = sub('(; [kpcofgs]__)+$', "", Taxon, perl=T)) %>%
  arrange(order(match(rownames(counts), `Feature ID`)))
### check if the order of the assignments and the order of feature table is the same
if (!all(rownames(counts) == ta$`Feature ID`)) {
  stop (simpleError("The order of the features in the table and classifications don't match"))
}
adf <- split_assignments(ta$trunc_taxon) 
### remove contamination
is_mitochondrial <- grepl("mitochondria", adf$Family)
is_chloroplast <- grepl("Chloroplast", adf$Class)
is_unassigned <- grepl("Unassigned", adf$Kingdom)
is_archaea <- grepl("Archaea", adf$Kingdom)
is_contam <- is_mitochondrial | is_chloroplast | is_unassigned ### Archaea kept to check positive control samples
counts <- counts[!is_contam,]
adf <- adf[!is_contam,]
ta <- ta[!is_contam,]
rm(is_contam, is_mitochondrial, is_chloroplast, is_unassigned, is_archaea)
a <- simplify_assignments(adf, rank1="Phylum", rank2="Genus")
names(a) <- ta$`Feature ID`
summed_cts <- rowsum(counts, a) 
summed_props <- sweep(summed_cts, 2, colSums(summed_cts), "/")
#if we want all the otus
 all_otus <- counts %>%
   merge(ta[,c("Feature ID","Confidence","trunc_taxon")], by.x="row.names", by.y="Feature ID", all.x=T) %>%
   dplyr::rename(Taxon = trunc_taxon) %>%
   dplyr::rename(`Feature ID` = Row.names)
# 
# write.csv(x = all_otus, file = "all_otus.csv", row.names = F)
```


```{r, Samples error check 1}
### ===========================
###   check for missing samples
### ===========================
### possible issue 1: Samples found in the sample sheet but not in the feature table (0 reads)
s_missing <- s %>%
  filter(!SampleID %in% colnames(counts)) %>%
  select(SampleID, SampleType, isControl)
if (any(!s_missing$isControl)) {
  pander(filter(s_missing, !isControl), caption="These samples were in the sample sheet but not in the feature table.")
}
```


```{r, Samples error check 2}
### possible issue 2: Samples found in the feature table but not in the sample sheet. There must be an error!
in_counts_not_in_s <- setdiff(colnames(counts), s$SampleID)
if (length(in_counts_not_in_s) > 0) {
  stop (simpleError("These SampleID(s) are in the feature table, but not found in the sample sheet.", paste(in_counts_not_in_s, collapse=" ")))
}
s[s$SampleID %in% s_missing$SampleID, "Keep"] <- FALSE
```


```{r theme, include=FALSE}
theme_set(theme_classic() + theme(
axis.line.x = element_line(colour = 'black', size=0.5, linetype='solid'),
axis.line.y = element_line(colour = 'black', size=0.5, linetype='solid'))) 

theme_boxplot <- function () {theme_base(base_size = 15) + 
                theme(line = element_line(size=0.8),
                text = element_text(size = 15),
                axis.text.x = element_text(color="black"),
                axis.text.y = element_text(color="black"),
                plot.background=element_blank())}

theme_figures <- function () {theme_kyle() +
                theme(line = element_line(size=0.8),
                text = element_text(size = 15),
                axis.text.x = element_text(color="black"),
                axis.text.y = element_text(color="black"),
                plot.background=element_blank())}
```


```{r general quantification, include=FALSE}
## Whole samples that are above the `r min_reads` read count threshold
s %>%
  mutate(KeepLabel = if_else(Keep, "Keep", "Discard")) %>%
  count(SampleType, KeepLabel) %>%
  spread(KeepLabel, n) %>%
  pander()

#Subset samples geneBlock_DNA
s <- filter(s, !is.na(HostSpecies))

```

```{r Color Palette, include=FALSE}
cols_group <- c("Control"= "#0000C0", "IONPs"= "#C0C000", "GOx" ="#A00000", "CHX" ="#76069A", "IONPs-GOx" ="#008000", "PBS" ="#ffac57", "NPC" ="#b70037", "NPC.Farnesol" ="#00b055", "Farnesol"= "#005ac1", "NPC.TB"= "#7cc9ff", "TB"= "#c667ef")

cols_Site <- c("Oral Swab"= "#6ad54a", "Dental Plaque"= "#4743c7", "Feces"= "#ae1c00")

```


```{r, warning=F}
### ===========================
###   calculate / read in alpha diversity
### ===========================
faith <- read.delim(file=faith_fp, stringsAsFactors = F) %>%
  dplyr::rename(SampleID=X)
# dplyr::filter(!isControl) %>%
s <- s %>%
  merge(vegan::diversity(t(counts)), by.x="SampleID", by.y="row.names", all.x=T) %>%
  dplyr::rename(shannon = y) %>%
  merge(rarefy(t(counts), richness_subsample_size), by.x="SampleID", by.y="row.names", all.x=T) %>%
  dplyr::rename(richness = y) %>%
  merge(faith, by="SampleID", all.x=T)
```

# Overview
This report is based on the results of 16S sequencing performed on MHK11.03.2019 for Yue's Project and MHK10.01.2019 for Tatsuro's Project.

Different treatments were performed in the 2 projects. Also, infection diferred:

1) MHK10.01.2019: _Streptococcus oralis_ and _S. mutans_

1) MHK11.03.2019: _Candida albicans_ and _S. mutans_

```{r, fig.width=6, fig.height=4, warning=F, include=FALSE}
## Histogram of high quality paired reads per sample
#The black dashed vertical line shows the minimum number of reads (`r min_reads`) for analysis. Control samples, if any, were included in the histogram.

ggplot(s, aes(x=Read_Counts)) +
    geom_histogram(aes(fill=SampleType), binwidth=1000) +
    geom_vline(xintercept = min_reads, color="black", linetype="dashed") +
    theme_classic() +
    theme_bw() + 
    xlab("Number of reads in sample") +
    ylab("Number of samples")
```

# Compare sites

```{r}
s_toPlot <- s %>%
  filter(Keep) %>%
  filter(Water=="Sucrose")
props_toPlot <- summed_props[, s_toPlot$SampleID]  

```

## Beta diversity

```{r}
dist_fp <- bc_fp
dist_name <- "Bray-Curtis Unifrac distances"
```
The `r dist_name` was employed to compare the species composition of the samples to each other. We used a method of ordination called Principal Coordinates Analysis to select the best 2D coordinate system for display. The percentage in each axis represents the proportion of the total variance captured along the axis. 
```{r Obtain the proportiona and axes for the amin genera}
taxon_props <- props_toPlot[rowMeans(props_toPlot) > 0.01,]
# Make sure the distance matrix is in the same order as the sample table
dist_in <- usedist::dist_subset(read_qiime_distmat(dist_fp), s_toPlot$SampleID)
# Here is the actual principal coordinates analysis
pc <- pcoa(dist_in)
# Get the first 3 principal coordinates as a matrix,
#   make sure rows are in the same order as sample table
s_coords <- pc$vectors[s_toPlot$SampleID,1:3]
# Get the taxon proportions as a matrix,
#   flip so that samples are in the rows,
#   again make sure rows are in the correct order
s_taxa <- t(taxon_props)[s_toPlot$SampleID,]

# Stick the coordinates and taxon proportions onto
#   the right side of the sample table
pc_df <- cbind(s_toPlot, s_coords, s_taxa)

```


```{r PCoA arrow}
# Make weights for each taxon
# Weights should sum to 1, divide by total for each column
taxon_weights <- apply(s_taxa, 2, function (x) x / sum(x))
# Matrix multiplication seems tricky here,
#   but we need to multiply sample weights by sample coordinates
#   and sum to get the average. We need to do this for each coordinate
#   axis and each taxon -- that's what matrix multiplication does.
#   You can try it by hand for one coordinate axis and taxon and
#   check that the answer is correct.
taxon_centers_of_gravity <- t(taxon_weights) %*% s_coords
# Convert to data frame
taxon_centers_of_gravity <- as.data.frame(taxon_centers_of_gravity)
# Dedicate one column to hold the taxon name
taxon_centers_of_gravity$Taxon <- rownames(taxon_centers_of_gravity)

# Splits the taxon name column by the 'space' into "Taxon" and "Taxon_name"
taxon_centers_of_gravity <- separate(taxon_centers_of_gravity, Taxon, sep=" ", c("Taxon","Taxon_name"))
# Removes the g__ or f__ in the Taxon_name column
taxon_centers_of_gravity$Taxon_name <- substring(taxon_centers_of_gravity$Taxon_name, 4)

#Percentaje of PC axes
pc <- pcoa(dist_in)
pctvar <- round(pc$values$Relative_eig * 100)

```

### Figure 1A

```{r wu PCoA PLOT}
pc_df %>%
  ggplot(aes(x=Axis.1, y=Axis.2)) +
  geom_point(
    aes(color=SampleType),
    shape= 21, fill="#ebcdc1",
    size = 2.5, stroke=2, alpha=0.8) +
  scale_color_manual(values = cols_Site) +
  geom_text_repel(
    aes(label=Taxon_name), fontface = 'italic',
    data=taxon_centers_of_gravity, segment.size=0) +
  stat_ellipse(aes(group=SampleType)) +
  labs(
    x=paste0("PC1 (", pctvar[1], "%)"),
    y=paste0("PC2 (", pctvar[2], "%)"),
    color="") +
  theme_boxplot()
```

```{r test, include=FALSE}
source("Function.Pairwise.Adonis.dm.R")
fstats_df <- pairwise.adonis.dm(dist_in, s_toPlot$SampleType)
```

### Fig 1D

Rat and human body-site differences

```{r}
fstats_df %>%
  ggplot() +
  geom_col(aes(x=pairs, y=R2, fill=pairs)) +
  scale_fill_simpsons(guide=FALSE) +
  scale_y_continuous(limits = c(0, 1)) +
  coord_flip() +
  labs(x="", y=parse(text='R^2')) +
  theme_boxplot()
```

## Alpha Diversity

```{r, fig.height=2.5}
s_toPlot %>% filter(Keep) %>%
  ggplot(aes(x=SampleType, y=richness)) +
  geom_boxplot() +
  geom_beeswarm(aes(color=SampleType), alpha=0.5) +
  scale_color_manual(values = cols_Site, guide=FALSE) +
  labs(x="", y="Richness\n(1k reads per sample)") +
  theme_boxplot()
```

```{r}
pairwise.wilcox.test(s_toPlot$richness, s_toPlot$SampleType, p.adjust.method = "bonferroni")
```



