---
title: "Body sites microbiome in animal model"
subtitle: "Figure 1B: Body sites HMP"
author: "Aurea Simon-Soro and Kyle Bittinger"
date: "06/12/2020"
output:
  pdf_document: default
  html_document: default
---


```{r knitr setup, echo=FALSE}
### ================
###   knitr setup
### ================
library(knitr)
opts_chunk$set(
  tidy=FALSE,
  cache=TRUE,
  echo=FALSE,
  warning=FALSE,
  message=FALSE,
  dpi=100,
  fig.width=8,
  fig.height=8,
  fig.align = "center"
  )
```

<!-- R packages -->
```{r R packages, message=FALSE}
### ================
###   R packages
### ================
#This package will also help us more easily manipulate our data
library(magrittr)
library(qiimer)
library(pander)
#Analyses of Phylogenetics and Evolution package. Required for tree calculations to be used with phyloseq
library(ape)
#The vegan package provides tools for descriptive community ecology. It has most basic functions of diversity analysis, community ordination and dissimilarity analysis. In general, this package is used for Bray-Curtis and Jaccard analyses.
library(vegan)
library(usedist)
library(tidyverse)
library(tibble)
#The phyloseq package seeks to address issues with multiple microbiome analysis packages by providing a set of functions that internally manage the organizing, linking, storing, and analyzing of phylogenetic sequencing data. In general, this package is used for UniFrac analyses.
library(phyloseq)
#Pretty Venn disgrams
library(VennDiagram)
library(ggbeeswarm)
library(ggtern)
library(ggthemes)
library(reshape2)
library(kylemisc)
library(ggrepel)
library(ggsci)
```



<!-- resources -->
```{r resources}
### ================
###   R resources
### ================
source("/home/aurea/R_functions.R")
```


```{r}
### ===========================
###   define constants
### ===========================
### minimum reads threshold
min_reads <- 1000
### rarefying subsample size 
richness_subsample_size <- 1000
### number of samples threshold to show heatmap on the page
sample_threshold <- 100
### setwd
#fill in your project dir
# From Kyle:
# To keep the project self-contained, place the data inside the project
# directory and use relative paths.  Here, I put the files in inst/extdata.
# If this project ever turns into an R package, you can keep the data files
# right where they are.
#Aurea: Here I used only the HMP study samples.
root_dir <- "/home/aurea/Koo/Shared.Species.BODYSITE/HMP/"
### mapping file path
mapping_file_fp <- file.path(root_dir, "MF_v13_16S.HMP.v4.txt")
### otu table file path
feature_table_fp <- file.path(root_dir, "feature-table.tsv")
### taxonomic assignment 
taxo_assignment_fp <- file.path(root_dir, "taxonomy.tsv")
### Bray-Curtis Dissimilarities file path
bc_fp <- file.path(root_dir, "bray_distance-matrix.tsv")
### faith
faith_fp <- file.path(root_dir, "alpha-diversity.tsv")
```

```{r}
study_group_levels <- c("Oral", "Gastrointestinal_tract")
sample_type_levels <- c("Buccal_mucosa", "Supragingival_plaque", "Stool")
```


```{r, warning=F, message=F}
### ===========================
###   read in data
### ===========================
### read mapping file
s <- read_tsv(mapping_file_fp) %>%
  rename(SampleID = "#SampleID") %>%
  # Fix sequencing control samples
  mutate(HMPBodySubsite = if_else(
    (HMPBodySite %in% "Contamination Control"), "Blank oral swab", HMPBodySubsite)) %>%
  mutate(HMPBodySite = na_if(HMPBodySite, "Contamination Control")) %>%
  # Assign levels
  mutate(HMPBodySite = fct_relevel(HMPBodySite, study_group_levels)) %>%
  #mutate(Water = fct_relevel(Water, water_levels)) %>%
  mutate(HMPBodySubsite = fct_relevel(HMPBodySubsite, sample_type_levels))
```


```{r message=FALSE}
### read otu table
assignments <- read_tsv(file=taxo_assignment_fp) %>%
  rename(Lineage = Taxon, OtuID = `Feature ID`) %>%
  do(bind_cols(., split_assignments(.$Lineage))) %>%
  mutate_at(taxonomic_ranks, str_remove, "[kpcofgs]__") %>%
  mutate_at(taxonomic_ranks, na_if, "") %>%
  mutate(Assignment = simplify_assignments(.[taxonomic_ranks]))
otu_counts <- readr::read_tsv(feature_table_fp, skip=1) %>%
  rename(OtuID = "#OTU ID") %>%
  gather(SampleID, Counts, -OtuID) %>%
  left_join(assignments, by="OtuID") %>%
  filter(!str_detect(Family, "mitochondria")) %>%
  filter(!str_detect(Class, "Chloroplast")) %>%
  filter(!str_detect(Kingdom, "Unassigned")) %>%
  filter(!str_detect(Kingdom, "Archaea")) %>%
  group_by(SampleID) %>%
  filter(sum(Counts) > 0)
if (!all(otu_counts$OtuID %in% assignments$OtuID)) {
  stop("Some OTUs not in taxonomic assignments table")
}
assignment_counts <- otu_counts %>%
  group_by(SampleID, Assignment) %>%
  summarize(Counts = sum(Counts)) %>%
  ungroup() %>%
  group_by(SampleID) %>%
  mutate(Proportion = Counts / sum(Counts)) %>%
  ungroup()
read_counts <- otu_counts %>%
  group_by(SampleID) %>%
  summarize(NumReads = sum(Counts))
```

No reads after contamination/archaea removal.

```{r, Samples error check 1}
### ===========================
###   check for missing samples
### ===========================
### possible issue 1: Samples found in the sample sheet but not in the feature table (0 reads)
s %>%
  filter(!(SampleID %in% read_counts$SampleID)) %>%
  select(SampleID, HMPBodySubsite)
```

Samples in feature table but not sample info table.

```{r, Samples error check 2}
### possible issue 2: Samples found in the feature table but not in the sample sheet. There must be an error!
read_counts %>%
  filter(!(SampleID %in% s$SampleID))
```


```{r theme, include=FALSE}
theme_set(theme_classic() + theme(
axis.line.x = element_line(colour = 'black', size=0.5, linetype='solid'),
axis.line.y = element_line(colour = 'black', size=0.5, linetype='solid'))) 
theme_boxplot <- function () {theme_base(base_size = 15) + 
                theme(line = element_line(size=0.8),
                text = element_text(size = 16),
                axis.text.x = element_text(color="black"),
                axis.text.y = element_text(color="black"),
                plot.background=element_blank())}
theme_figures <- function () {theme_kyle() +
                theme(line = element_line(size=0.8),
                text = element_text(size = 16),
                axis.text.x = element_text(color="black"),
                axis.text.y = element_text(color="black"),
                plot.background=element_blank())}
theme_scatterplot <- function () {theme_base(base_size = 12) + theme(
  line = element_line(size=0.8),
  text = element_text(face = "bold", color = "black"),
  axis.text.x = element_text(face="bold", color="black"),
  axis.text.y = element_text(face="bold", color="black"),
                plot.background=element_blank())}
```



```{r general quantification, include=FALSE}
## Whole samples that are above the `r min_reads` read count threshold
###Keep only the samples read counts > 1000 seqs
s %>%
  left_join(read_counts, by="SampleID") %>%
  mutate(NumReads = replace_na(NumReads, 0)) %>%
  mutate(Keep = if_else(NumReads > min_reads, "Keep", "Discard")) %>%
  count(HMPBodySubsite, Keep) %>%
  spread(Keep, n) %>%
  pander()
```

```{r Color Palette, include=FALSE}
cols_Site <- c("#6ad54a", "#4743c7", "#ae1c00")
```


```{r}
# read in beta diversity
bc <- read_qiime_distmat(bc_fp)
```

# Compare sites

```{r}
s_sites <- s %>%
  left_join(read_counts, by="SampleID") %>%
  filter(NumReads > min_reads)
```

##Figure1

```{r}
dist_fp <- bc_fp
dist_name <- "Bray-Curtis Unifrac distances"
```
The `r dist_name` was employed to compare the species composition of the samples to each other. We used a method of ordination called Principal Coordinates Analysis to select the best 2D coordinate system for display. The percentage in each axis represents the proportion of the total variance captured along the axis. 
```{r Obtain the proportion and axes for the main genera}
bc_sites <- dist_subset(bc, s_sites$SampleID)
# Here is the actual principal coordinates analysis
pc <- pcoa(bc_sites)
pctvar <- round(pc$values$Relative_eig * 100)
# Principal coordinate positions of samples
s_coords <- pc$vectors %>%
  as.data.frame() %>%
  rownames_to_column("SampleID") %>%
  select(1:4)
# Principal coordinate positions of taxa
taxa_coords <- assignment_counts %>%
  # Filter to include samples in this analysis
  right_join(s_coords, by="SampleID") %>%
  # Filter to keep high-abundance taxa
  group_by(Assignment) %>%
  filter(mean(Proportion) > 0.02) %>%
  ungroup() %>%
  # Weighted mean position along each axis
  gather(Axis, SamplePosition, starts_with("Axis")) %>%
  group_by(Assignment, Axis) %>%
  summarize(
    TaxonPosition = weighted.mean(SamplePosition, Proportion),
    MeanProp = mean(Proportion)) %>%
  ungroup() %>%
  spread(Axis, TaxonPosition) %>%
  # Make the labels nicer
  mutate(TaxonLabel = str_remove(Assignment, "^\\w+ "))
```

###Fig 1C
```{r fig.width=8, fig.height=5.5}
#When I want to rotate the plot and the names of the plot as well
taxa_coords <- taxa_coords %>%
mutate(Axis.1=-Axis.1) %>%
mutate(Axis.2=-Axis.2)

s_sites %>%
  left_join(s_coords, by="SampleID") %>%
  mutate(Axis.1 = -Axis.1) %>% 
  mutate(Axis.2 = -Axis.2) %>%
  ggplot(aes(x=Axis.1, y=Axis.2)) +
  geom_point(
    aes(fill=HMPBodySubsite), shape= 21, size=5) +
  scale_fill_manual(values = cols_Site) +
  #geom_text_repel(
    #aes(label=TaxonLabel), fontface = 'bold.italic',
    #data=taxa_coords, segment.size=0) +
  stat_ellipse(aes(group=HMPBodySubsite)) +
  labs(
    x=paste0("PC1 (", pctvar[1], "%)"),
    y=paste0("PC2 (", pctvar[2], "%)"),
     fill="") +
   theme_scatterplot() +
 #theme(legend.position="bottom") +
  theme(text = element_text(size = 16)) +
  xlim(-0.7, 0.7) +
  ylim(-0.7, 0.7)
```


```{r test, include=FALSE}
source("/home/aurea/Function.Pairwise.Adonis.dm.R")
fstats_df <- pairwise.adonis.dm(bc_sites, s_sites$HMPBodySubsite)
fstats_df
```

### Fig 1D (lower panel: human)
Rat and human body-site differences
```{r fig.height=2, fig.width=6}
fstats_df %>%
  ggplot() +
  geom_col(aes(x=pairs, y=R2, fill=pairs)) +
  scale_fill_manual(values = cols_Site, guide=FALSE) +
  scale_y_continuous(limits = c(0, 1)) +
  coord_flip() +
  labs(x="", y=parse(text='R^2')) +
  theme_boxplot()
```




