---
title: "Report: Body-sites microbiome in animal model"
subtitle: "Figure 1: Body-sites microbiome description"
author: "Aurea Simon-Soro"
date: \today
output:
  pdf_document: default
  html_document: default
---

```{r knitr setup, echo=FALSE}
knitr::opts_chunk$set(
  tidy=FALSE,
  cache=TRUE,
  echo=FALSE,
  warning=FALSE,
  message=FALSE,
  dpi=100,
  fig.width=8,
  fig.height=8,
  fig.align = "center"
  )
```

## Setup

```{r}
library(tidyverse)
library(qiimer)
library(vegan)
```

```{r}
MIN_READS <- 1000
```


```{r theme, include=FALSE}
theme_set(theme_classic() + theme(
axis.line.x = element_line(colour = 'black', size=0.5, linetype='solid'),
axis.line.y = element_line(colour = 'black', size=0.5, linetype='solid')))

theme_boxplot <- function () {
  ggthemes::theme_base(base_size = 15) +
  theme(
    line = element_line(size=0.8),
    text = element_text(size = 15),
    axis.text.x = element_text(color="black"),
    axis.text.y = element_text(color="black"),
    plot.background=element_blank())
}

theme_figures <- function () {
  kylemisc::theme_kyle() +
  theme(
    line = element_line(size=0.8),
    text = element_text(size = 15),
    axis.text.x = element_text(color="black"),
    axis.text.y = element_text(color="black"),
    plot.background=element_blank())
}
```

```{r}
study_group_levels <- c(
  "Control", "IONPs", "GOx", "CHX",
  "IONPs-GOx", "PBS", "NPC", "NPC.Farnesol",
  "Farnesol", "NPC.TB", "TB")
water_levels <- c("Sucrose", "Glucose")
sample_type_levels <- c("Oral Swab", "Dental Plaque", "Feces")
```

```{r Color Palette, include=FALSE}
cols_group <- c("Control"= "#0000C0", "IONPs"= "#C0C000", "GOx" ="#A00000", "CHX" ="#76069A", "IONPs-GOx" ="#008000", "PBS" ="#ffac57", "NPC" ="#b70037", "NPC.Farnesol" ="#00b055", "Farnesol"= "#005ac1", "NPC.TB"= "#7cc9ff", "TB"= "#c667ef")

cols_Site <- c("#6ad54a", "#4743c7", "#ae1c00")
```

## Load data

```{r, warning=F, message=F}
s <- read_tsv("inst/extdata/MF.Aurea.txt") %>%
  rename(SampleID = "#SampleID") %>%

  # Fix sequencing control samples
  mutate(SampleType = if_else(
    (study_group %in% "Contamination Control"), "Blank oral swab", SampleType)) %>%
  mutate(is_seq_control = is.na(HostSpecies)) %>%
  mutate(study_color = na_if(study_color, "Contamination Control")) %>%
  mutate(study_group = na_if(study_group, "Contamination Control")) %>%

  # Assign levels
  mutate(study_group = fct_relevel(study_group, study_group_levels)) %>%
  mutate(Water = fct_relevel(Water, water_levels)) %>%
  mutate(SampleType = fct_relevel(SampleType, sample_type_levels))
```


```{r message=FALSE}
assignments <- read_tsv("inst/extdata/taxonomy.tsv") %>%
  rename(Lineage = Taxon, OtuID = `Feature ID`) %>%
  do(bind_cols(., split_assignments(.$Lineage))) %>%
  mutate_at(taxonomic_ranks, str_remove, "[kpcofgs]__") %>%
  mutate_at(taxonomic_ranks, na_if, "") %>%
  mutate(Assignment = simplify_assignments(.[taxonomic_ranks]))

otu_counts <- readr::read_tsv("inst/extdata/feature-table.tsv", skip=1) %>%
  rename(OtuID = "#OTU ID") %>%
  gather(SampleID, Counts, -OtuID) %>%
  left_join(assignments, by="OtuID") %>%
  filter(!str_detect(Family, "mitochondria")) %>%
  filter(!str_detect(Class, "Chloroplast")) %>%
  filter(!str_detect(Kingdom, "Unassigned")) %>%
  filter(!str_detect(Kingdom, "Archaea")) %>%
  group_by(SampleID) %>%
  filter(sum(Counts) > 0)

if (!all(otu_counts$OtuID %in% assignments$OtuID)) {
  stop("Some OTUs not in taxonomic assignments table")
}

assignment_counts <- otu_counts %>%
  group_by(SampleID, Assignment) %>%
  summarize(Counts = sum(Counts)) %>%
  ungroup() %>%
  group_by(SampleID) %>%
  mutate(Proportion = Counts / sum(Counts)) %>%
  ungroup()

read_counts <- otu_counts %>%
  group_by(SampleID) %>%
  summarize(NumReads = sum(Counts))
```

```{r message=FALSE, warning=FALSE}
faith_df <- read_tsv("inst/extdata/alpha-diversity.tsv") %>%
  rename(SampleID=X1)

alpha_df <- otu_counts %>%
  group_by(SampleID) %>%
  summarize(
    Shannon = diversity(Counts),
    Richness = rarefy(Counts, 1000)) %>%
  left_join(faith_df, by="SampleID")
```

```{r}
bc <- read_qiime_distmat("inst/extdata/bc.distance-matrix.tsv")
```


## Quality control

Some samples have no reads after contamination/archaea removal. Samples in the sample info table but not in the feature table:

```{r}
s %>%
  filter(!(SampleID %in% read_counts$SampleID)) %>%
  select(SampleID, SampleType)
```

Samples in feature table but not sample info table:

```{r, Samples error check 2}
read_counts %>%
  filter(!(SampleID %in% s$SampleID))
```

```{r}
s %>%
  left_join(read_counts, by="SampleID") %>%
  mutate(NumReads = replace_na(NumReads, 0)) %>%
  mutate(Keep = if_else(NumReads > MIN_READS, "Keep", "Discard")) %>%
  count(SampleType, Keep) %>%
  spread(Keep, n, fill = 0)
```




# Overview

This report is based on the results of 16S sequencing performed on MHK11.03.2019 for Yue's Project and MHK10.01.2019 for Tatsuro's Project.

Different treatments were performed in the 2 projects. Also, infection diferred:

1) MHK10.01.2019: _Streptococcus oralis_ and _S. mutans_

1) MHK11.03.2019: _Candida albicans_ and _S. mutans_

```{r, fig.width=6, fig.height=4, warning=F}
s %>%
  left_join(read_counts, by="SampleID") %>%
  ggplot(aes(x=NumReads)) +
  geom_histogram(aes(fill=SampleType), binwidth=1000) +
  geom_vline(xintercept = MIN_READS, color="black", linetype="dashed") +
  ggsci::scale_fill_d3() +
  labs(x="Number of reads", y="Number of samples") +
  theme_bw()
```

# Compare sites

```{r}
s_sites <- s %>%
  left_join(read_counts, by="SampleID") %>%
  filter(NumReads > MIN_READS) %>%
  filter(Water %in% "Sucrose")
```

## Beta diversity

```{r}
dist_name <- "Bray-Curtis Unifrac distances"
```
The `r dist_name` was employed to compare the species composition of the samples to each other. We used a method of ordination called Principal Coordinates Analysis to select the best 2D coordinate system for display. The percentage in each axis represents the proportion of the total variance captured along the axis.
```{r Obtain the proportiona and axes for the amin genera}
bc_sites <- dist_subset(bc, s_sites$SampleID)

# Here is the actual principal coordinates analysis
pc <- ape::pcoa(bc_sites)

pctvar <- round(pc$values$Relative_eig * 100)

# Principal coordinate positions of samples
s_coords <- pc$vectors %>%
  as.data.frame() %>%
  rownames_to_column("SampleID") %>%
  select(1:4)

# Principal coordinate positions of taxa
taxa_coords <- assignment_counts %>%
  # Filter to include samples in this analysis
  right_join(s_coords, by="SampleID") %>%
  # Filter to keep high-abundance taxa
  group_by(Assignment) %>%
  filter(mean(Proportion) > 0.02) %>%
  ungroup() %>%
  # Weighted mean position along each axis
  gather(Axis, SamplePosition, starts_with("Axis")) %>%
  group_by(Assignment, Axis) %>%
  summarize(
    TaxonPosition = weighted.mean(SamplePosition, Proportion),
    MeanProp = mean(Proportion)) %>%
  ungroup() %>%
  spread(Axis, TaxonPosition) %>%
  # Make the labels nicer
  mutate(TaxonLabel = str_remove(Assignment, "^\\w+ "))
```

### Figure 1A

```{r fig.width=7, fig.height=5}
s_sites %>%
  left_join(s_coords, by="SampleID") %>%
  ggplot(aes(x=Axis.1, y=Axis.2)) +
  geom_point(
    aes(color=SampleType),
    shape= 21, fill="#ebcdc1",
    size = 2.2, stroke=2, alpha=0.7) +
  scale_color_manual(values = cols_Site) +
  ggrepel::geom_text_repel(
    aes(label=TaxonLabel), fontface = 'italic',
    data=taxa_coords, segment.size=0) +
  stat_ellipse(aes(group=SampleType)) +
  labs(
    x=paste0("PC1 (", pctvar[1], "%)"),
    y=paste0("PC2 (", pctvar[2], "%)"),
    color="") +
  theme_boxplot()
ggsave("inst/figures/sites_braycurtis.pdf", width=7, height=5, useDingbats=F)
```

```{r test, include=FALSE}
source("Function.Pairwise.Adonis.dm.R")
fstats_df <- pairwise.adonis.dm(bc_sites, s_sites$SampleType)
fstats_df
```

### Fig 1D

Rat and human body-site differences

```{r fig.height=2, fig.width=5}
fstats_df %>%
  ggplot() +
  geom_col(aes(x=pairs, y=R2, fill=pairs)) +
  ggsci::scale_fill_simpsons(guide=FALSE) +
  scale_y_continuous(limits = c(0, 1)) +
  coord_flip() +
  labs(x="", y=parse(text='R^2')) +
  theme_boxplot()
ggsave("inst/figures/sites_r2.pdf", width=5, height=2)
```

## Alpha Diversity

```{r, fig.height=3, fig.width=4}
s_sites %>%
  mutate(SampleType = str_replace(SampleType, " ", "\n")) %>%
  mutate(SampleType = fct_relevel(SampleType, "Oral\nSwab")) %>%
  left_join(alpha_df, by="SampleID") %>%
  ggplot(aes(x=SampleType, y=Richness)) +
  geom_boxplot(outlier.shape = NA) +
  ggbeeswarm::geom_beeswarm(aes(color=SampleType), alpha=0.5) +
  scale_color_manual(values = cols_Site, guide=FALSE) +
  labs(x="", y="Richness\n(1k reads per sample)") +
  theme_boxplot()
ggsave("inst/figures/sites_richness.pdf", width=4, height=3, useDingbats=F)
```

```{r}
s_sites %>%
  left_join(alpha_df, by="SampleID") %>%
  with(pairwise.wilcox.test(
    Richness, SampleType, p.adjust.method = "bonferroni"))
```

# Nestedness and turnover

```{r}
s_nest <- s %>%
  filter(project %in% "MHK10.01.2019") %>%
  filter(!is.na(study_group)) %>%
  filter(!str_detect(study_group, "TB")) %>%
  droplevels()
```

```{r}
s_nest %>%
  xtabs(~ study_group + SampleType, data = .)
```

```{r}
library(betapart)
pres_nest <- otu_counts %>%
  mutate(Present = as.numeric(Counts > 0)) %>%
  right_join(s_nest, by="SampleID") %>%
  group_by(OtuID) %>%
  #filter(sum(Present) > 0) %>%
  ungroup() %>%
  usedist::pivot_to_numeric_matrix(SampleID, OtuID, Present)
jacc_nest <- beta.pair(pres_nest, index.family = "jaccard")
```

```{r}
jne_df <- jacc_nest$beta.jne %>%
  usedist::dist_subset(s_nest$SampleID) %>%
  usedist::dist_groups(s_nest$SampleType) %>%
  mutate(Component = "Nestedness")

jtu_df <- jacc_nest$beta.jtu %>%
  usedist::dist_subset(s_nest$SampleID) %>%
  usedist::dist_groups(s_nest$SampleType) %>%
  mutate(Component = "Turnover")

jac_df <- jacc_nest$beta.jac %>%
  usedist::dist_subset(s_nest$SampleID) %>%
  usedist::dist_groups(s_nest$SampleType) %>%
  mutate(Component = "Total")

jacc_nest_df <- bind_rows(jne_df, jtu_df) %>%
  arrange(Item1, Item2, Component) %>%
  group_by(Item1, Item2) %>%
  mutate(TotalDistance = sum(Distance)) %>%
  mutate(FractDistance = Distance / TotalDistance) %>%
  ungroup()

sample_group_levels <- c(
  "Within Oral Swab",
  "Within Dental Plaque",
  "Within Feces",
  "Between Oral Swab and Dental Plaque",
  "Between Dental Plaque and Feces",
  "Between Oral Swab and Feces")
```


```{r}
bind_rows(jne_df, jtu_df, jac_df) %>%
  mutate(Label = fct_relevel(Label, rev(sample_group_levels))) %>%
  mutate(Component = fct_relevel(Component, "Nestedness", "Turnover")) %>%
  mutate(Component = fct_recode(
    Component,
    "Nestedness component" = "Nestedness",
    "Turnover component" = "Turnover",
    "Total Jaccard distance" = "Total")) %>%
  ggplot() +
  geom_boxplot(aes(x=Label, y=Distance)) +
  facet_grid(~ Component) +
  coord_flip() +
  scale_y_continuous(breaks = c(seq(0, 1, by=0.5))) +
  labs(x="", y="", title="Components of Jaccard distance") +
  theme(strip.background = element_blank())
ggsave("inst/figures/sites_nestedness.pdf", width=10, height=4, useDingbats=F)
```

```{r}
bind_rows(jne_df, jtu_df) %>%
  pivot_wider(names_from = Component, values_from = Distance) %>%
  mutate(Jaccard = Nestedness + Turnover) %>%
  mutate(FracTurnover = Turnover / Jaccard) %>%
  ggplot() +
  geom_point(aes(x=Jaccard, y=Turnover, color=Label)) +
  facet_wrap(~ Label)
```


```{r}
jacc_nest_df %>%
  mutate(Label = fct_relevel(Label, rev(sample_group_levels))) %>%
  ggplot() +
  geom_boxplot(aes(x=Label, y=FractDistance)) +
  facet_grid(~ Component) +
  coord_flip() +
  labs(x="", title="Fractional components of Jaccard distance")
ggsave("inst/figures/sites_nestedness_frac.pdf", width=8, height=4, useDingbats=F)
```

# Shared species

## Oral swab - plaque

```{r}
subj_pairs <- s_sites %>%
  filter(!(SampleType %in% "Feces")) %>%
  pivot_wider(
    id_cols = SubjectID,
    names_from = SampleType,
    values_from = SampleID) %>%
  filter(!is.na(`Dental Plaque`), !is.na(`Oral Swab`)) %>%
  pivot_longer(
    -SubjectID, names_to = "SampleType", values_to = "SampleID") %>%
  select(SampleID)
```

```{r}
last_word <- function (x) {
  x %>%
    str_split(" ") %>%
    vapply(function (y) y[length(y)], "hello")
}
paired_props <- s_sites %>%
  right_join(subj_pairs, by="SampleID") %>%
  left_join(otu_counts, by="SampleID") %>%
  group_by(SampleID) %>%
  mutate(Proportion = Counts / sum(Counts)) %>%
  ungroup() %>%
  select(SubjectID, SampleType, OtuID, Assignment, Proportion, project, study_group) %>%
  pivot_wider(
    names_from = SampleType, values_from = Proportion) %>%
  filter(((`Dental Plaque` > 0) | (`Oral Swab` > 0))) %>%
  filter(project %in% "MHK10.01.2019") %>%
  filter(!is.na(study_group)) %>%
  filter(!str_detect(study_group, "TB")) %>%
  mutate(AssignmentLabel = fct_lump(
    last_word(Assignment), n=12, w=`Oral Swab`)) %>%
  filter(!(AssignmentLabel %in% "Other"))
```


```{r}
paired_colors <- rcartocolor::carto_pal(12, "Vivid")
paired_props %>%
  mutate(`Oral Swab` = `Oral Swab` + 1e-5) %>%
  mutate(`Dental Plaque` = `Dental Plaque` + 1e-5) %>%
  ggplot() +
  geom_point(
    aes(x=`Oral Swab`, y=`Dental Plaque`, color = AssignmentLabel)) +
  geom_abline(slope = 1, intercept = 0, linetype = "dashed") +
  facet_wrap(~ AssignmentLabel, ncol = 4) +
  scale_color_manual(
    values = paired_colors, guide=FALSE) +
  scale_x_log10(
    labels = scales::trans_format("log10", scales::math_format(10^.x))) +
  scale_y_log10(
    labels = scales::trans_format("log10", scales::math_format(10^.x))) +
  labs(
    x = "Abundance in paired oral swab sample",
    y = "Abundance in paired dental plaque sample") +
  theme_bw() +
  theme(strip.background = element_blank())
ggsave(
  "inst/figures/sites_taxon_props.pdf",
  width=8, height=6, useDingbats=F)
```

```{r}
paired_props %>%
  filter(`Oral Swab` > 0) %>%
  mutate(PlaquePresent = factor(ifelse(`Dental Plaque` > 0, "Present", "Absent"))) %>%
  mutate(LogSwab = log10(`Oral Swab`)) %>%
  group_by(AssignmentLabel) %>%
  do(broom::tidy(glm(PlaquePresent ~ LogSwab, family = "binomial", data=.))) %>%
  mutate(estimate = estimate / log(10)) %>%
  mutate(
    sig = p.value < 0.05,
    estimate_plus = estimate + std.error,
    estimate_minus = estimate - std.error) %>%
  filter(sig) %>%
  ggplot(aes(x=AssignmentLabel, y=estimate, shape=sig)) +
  geom_point() +
  facet_grid(~ term)
```


```{r}
paired_props %>%
  filter(`Oral Swab` > 0) %>%
  mutate(PlaquePresent = as.numeric(`Dental Plaque` > 0)) %>%
  ggplot(aes(x=`Oral Swab`, y=PlaquePresent)) +
  geom_point() +
  geom_smooth(se = FALSE, method="glm", method.args=list(family="binomial")) +
  scale_x_log10() +
  facet_wrap(~ AssignmentLabel)
```

```{r}
paired_props %>%
  filter(`Dental Plaque` > 0) %>%
  mutate(SwabPresent = as.numeric(`Oral Swab` > 0)) %>%
  ggplot(aes(x=`Dental Plaque`, y=SwabPresent)) +
  geom_point() +
  geom_smooth(se = FALSE, method="lm", formula = y ~ 1) +
  scale_x_log10() +
  facet_wrap(~ AssignmentLabel)
```

```{r}
paired_props %>%
  rename(Oral = `Oral Swab`, Plaque = `Dental Plaque`) %>%
  mutate(MeanProp = (Oral + Plaque) / 2) %>%
  mutate(OralPresent = Oral > 0, PlaquePresent = Plaque > 0) %>%
  mutate(BothPresent = OralPresent & PlaquePresent) %>%
  mutate(OnePresent = xor(OralPresent, PlaquePresent)) %>%
  ggplot(aes(x=MeanProp, y=as.numeric(BothPresent))) +
  geom_point() +
  geom_smooth(se = FALSE, method="glm", method.args=list(family="binomial")) +
  scale_x_log10() +
  facet_wrap(~ AssignmentLabel)
```


Probability of an OTU appearing in oral swab also being present in dental
plaque.

```{r}
oral_asvs <- paired_props %>%
  mutate(AssignmentLabel = fct_lump(Assignment, n=10, w=`Oral Swab`)) %>%
  #filter(!(AssignmentLabel %in% "Other")) %>%
  filter(`Oral Swab` > 0) %>%
  mutate(PlaquePresent = `Dental Plaque` > 0)
```

```{r}
oral_asvs %>%
  ggplot() +
  geom_bar(aes(x = SubjectID, fill = PlaquePresent)) +
  facet_wrap(~ AssignmentLabel) +
  theme_bw() +
  theme(axis.text.x = element_blank())
```

```{r eval=FALSE}
library(lme4)
library(lmerTest)
oral_asvs %>%
  mutate(PlaquePres = `Dental Plaque` > 0) %>%
  mutate(LogOral = log(`Oral Swab`)) %>%
  group_by(AssignmentLabel) %>%
  do(glmer(PlaquePres ~ LogOral + (1|SubjectID), data=., family="binomial"))
  group_by(SubjectID, AssignmentLabel) %>%
  summarize(plaque_prob = mean(`Dental Plaque` > 0), n = n()) %>%
  ggplot() +
  geom_point(aes(x=plaque_prob, y = n, color=AssignmentLabel))
```



```{r}
library(polyafit)
library(usedist)

cts_matrix <- s_sites %>%
  right_join(subj_pairs, by="SampleID") %>%
  left_join(otu_counts, by="SampleID") %>%
  usedist::pivot_to_numeric_matrix(SampleID, OtuID, Counts)

get_matrix <- function (sample_ids, min_colsum = 5) {
  mat <- cts_matrix[sample_ids,]
  keep_col <- colSums(mat) > min_colsum
  mat <- mat[,keep_col]
}

fit_polya <- function (sample_ids, min_colsum = 5) {
  m <- get_matrix(sample_ids, min_colsum)
  pars <- polyafit::optim_polya(m)$par
  res <- apply(m, 1, ppolya_marginal, pars, log.p = FALSE)
  rownames(res) <- names(pars)
  as.data.frame(res) %>%
    rownames_to_column("OtuID") %>%
    pivot_longer(-OtuID, names_to = "SampleID", values_to = "polya_pval")
}



paired_cts <- s_sites %>%
  right_join(subj_pairs, by="SampleID") %>%
  group_by(SubjectID) %>%
  do(fit_polya(.$SampleID))


sig_asvs <- paired_cts %>%
  group_by(SubjectID, OtuID) %>%
  summarise(pval = min(polya_pval)) %>%
  ungroup() %>%
  mutate(plabel = cut(
    pval, breaks = c(0, 0.001, 0.01, 0.05, 1),
    labels = c("P < 0.001", "P < 0.01", "P < 0.05", "n.s."),
    include.lowest = TRUE))
```



# TB vs PBS

```{r}
s_tb <- s %>%
  left_join(read_counts, by="SampleID") %>%
  filter(NumReads > MIN_READS) %>%
  filter(Water %in% "Sucrose", study_group %in% c("PBS", "TB"))
```


```{r}
tb_taxa <- assignment_counts %>%
  filter(SampleID %in% s_tb$SampleID) %>%
  mutate(Assignment = fct_lump(Assignment, n=10, w=Proportion)) %>%
  # Sum in each sample
  group_by(SampleID, Assignment) %>%
  summarize(Proportion = sum(Proportion)) %>%
  ungroup() %>%
  # Average within groups
  left_join(s_tb, by="SampleID") %>%
  group_by(SampleType, study_group, Assignment) %>%
  summarize(Proportion = mean(Proportion)) %>%
  ungroup()
```

## Stacked bars

Bacteria with more than 1% mean relative abundance

```{r Bar Plot Taxa, fig.height=5}
tb_taxa %>%
  ggplot(aes(x=study_group, y=Proportion)) +
  geom_col(aes(fill = Assignment)) +
  labs(x= "Treatment", y= "Bacterial Abundance") +
  ggsci::scale_fill_d3(palette = c("category20")) +
  labs(x="", y="Relative Abundance", fill="Bacteria") +
  theme_boxplot() +
  facet_grid(~SampleType, scales= "free") +
  theme(axis.text.x=element_text(angle=-50, hjust= .1))
ggsave("inst/figures/tb_stackedbars.pdf", width=8, height=5)
```

## Beta diversity

```{r}
make_body_site_pcoa <- function (sample_type) {
  s_temp <- s_tb %>%
    filter(SampleType %in% sample_type)
  bc_temp <- dist_subset(bc, s_temp$SampleID)
  pc <- ape::pcoa(bc_temp)
  pctvar <- round(pc$values$Relative_eig * 100)
  coords <- pc$vectors %>%
    as.data.frame() %>%
    rownames_to_column("SampleID") %>%
    select(1:4)
  s_temp %>%
    left_join(coords, by="SampleID") %>%
    ggplot(aes(x=Axis.1, y=Axis.2, color=study_group)) +
    geom_point(, shape= 21, fill="#ebcdc1", size = 2.5, stroke=2, alpha=0.8) +
    scale_color_manual(values= cols_group) +
    labs(
      x=paste0("PC1 (", pctvar[1], "%)"),
      y=paste0("PC2 (", pctvar[2], "%)"),
      color="Treatment") +
    ggtitle(sample_type) +
    theme_boxplot() +
    theme(plot.title = element_text(hjust = 0.5))
}
```

```{r fig.height=3, fig.width=8}
p1 <- make_body_site_pcoa("Oral Swab")
p2 <- make_body_site_pcoa("Dental Plaque")
p3 <- make_body_site_pcoa("Feces")
ggpubr::ggarrange(p1, p2, p3, ncol = 3, nrow = 1, common.legend = TRUE)
ggsave("inst/figures/tb_braycurtis.pdf", width=8, height=3, useDingbats=F)
```

```{r}
test_body_site_dist <- function (sample_type) {
  s_temp <- s_tb %>%
    filter(SampleType %in% sample_type)
  bc_temp <- dist_subset(bc, s_temp$SampleID)
  res <- adonis(bc_temp ~ study_group, data=s_temp)
  res$aov.tab %>%
    as_tibble(rownames="Term")
}
tibble(body_site = sample_type_levels) %>%
  group_by(body_site) %>%
  do(test_body_site_dist(.$body_site)) %>%
  filter(Term %in% "study_group") %>%
  rename(Pvalue = `Pr(>F)`) %>%
  select(Term, R2, Pvalue)
```


## Figure 4

4B: correlation networks
oral, dental, fecal
color correlations to show disruption by TB

4C: LDA
build linear discriminants for oral, dental, fecal
use oral discriminant to predict TB effect in dental, fecal
